### DISCUSSION TITLE
Handling reserved names

### DISCUSSION THREAD

**Discussion by rafalkrupinski** [python-lapidary/lapidary#22] [Ideas]
> Talking about additionalProperties: https://github.com/oeklo/gsmtasks-client/blob/main/gen/gsmtasks/components/schemas/account.py#L24:
> 
> ```python
> import pydantic
> 
> class AccountTaskExpiryDurationFromCompleteAfter(pydantic.BaseModel):
>     class Config(pydantic.BaseConfig):
>         extra = pydantic.Extra.allow
> 
> b = AccountTaskExpiryDurationFromCompleteAfter(dict='yes')
> b.dict()
> #Traceback (most recent call last):
> #  File "<stdin>", line 1, in <module>
> #TypeError: 'str' object is not callable
> ```

It's a known pydantic limitation and they address it in v2 by using longer names :)

Anything but a pure dict would have some kind of limitation.
```python
class MyModel:
    def __init__(self, **kwargs):
        self.__dict__ = kwargs

m=MyModel(__dict__='yes')
print(m.__dict__)
# {'__dict__': 'yes'}
```

One solution would be to use TypedDict, but I'm not too fond of the syntax.

**Comment by commonism** [comment]
> It's a known pydantic limitation and they address it in v2 by using longer names :)

This still allows breaking the client using special input - I do not consider this a solution.

I'd prefer 
```python
class AccountTaskExpiryDurationFromCompleteAfter(pydantic.BaseModel):
    class Config(pydantic.BaseConfig):
        extra = pydantic.Extra.dict

b = AccountTaskExpiryDurationFromCompleteAfter(dict='yes')
b.__dict__["dict"]
```
Which does exactly whats required, accept everything, exposes/overwrites/breaks nothing

**Reply by rafalkrupinski** [reply]
I'm not sure if I follow the idea behind this example.
You are aware that `__dict__` is special in python, right?

**Comment by commonism** [comment]
You are right, using __dict__ is opposing the actual idea.

**Comment by rafalkrupinski** [comment]
Take a look how pydantic solves it. They put all data in a dict called `__fields_set__`.
Users access it thru `__getattr__` interface, which makes precedence for fields over reserved methods. That's why your dict field overrides BaseModel.dict method. It's their design choice.

I think the BaseModel methods should take precedence, and users could access fields with reserved names explicitly via `__field_set__`.

Other options are field alias for explicit properties, model-global or library-global alias generators. But it would require good name mangling to avoid name clashes. The one I have now (replacing all special characters with underscore) is not enough.

**Comment by commonism** [comment]
__fields_set__ = set(values.keys())
it is just the names of the fields which were set - not including default values
the data is stored in self.__dict__
https://github.com/pydantic/pydantic/blob/e34ff9255f281eaccac10d0436c082626d9146c3/pydantic/main.py#L578

storing the data in self.__values__: Dict and using a ChainMap(self.__dict__, self.__values__) in __getattr__ … sounds easy, but is a really hard cut into the pydantic DNA.

want to bug @samuelcolvin about it?

**Reply by samuelcolvin** [reply]
Humm, I'll think about this, v2 would be the time to solve it.

Please can someone create a feature request issue on pydantic to remind me about with a link to here.

**Reply by commonism** [reply]
https://github.com/pydantic/pydantic/issues/4617
